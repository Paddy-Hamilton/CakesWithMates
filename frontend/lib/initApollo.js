import { ApolloClient, InMemoryCache, HttpLink } from "apollo-boost";
import fetch from "isomorphic-unfetch";
import { EDIT_CAKE_MODAL_OPEN } from "../graphql/queries.graphql";
let apolloClient = null;

function create(initialState) {
  // Check out https://github.com/zeit/next.js/pull/4611 if you want to use the AWSAppSyncClient
  const isBrowser = typeof window !== "undefined";
  const cache = new InMemoryCache().restore(initialState || {});
  const client = new ApolloClient({
    connectToDevTools: isBrowser,
    ssrMode: !isBrowser, // Disables forceFetch on the server (so queries are only run once)
    link: new HttpLink({
      uri:
        process.env.NODE_ENV === "development"
          ? "http://localhost:8000/graphql"
          : "https://ph-cake-api.now.sh/graphql", // Server URL (must be absolute)
      credentials: "same-origin", // Additional fetch() options like `credentials` or `headers`
      // Use fetch() polyfill on the server
      fetch: !isBrowser && fetch
    }),
    cache,
    request: async operation => {
      operation.setContext({
        fetchOptions: {
          credentials: "include"
        },
        headers
      });
    },

    resolvers: {
      Mutation: {
        toggleEditCakeModal(_, variables, { cache }) {
          const { editCakeModalOpen } = cache.readQuery({
            query: EDIT_CAKE_MODAL_OPEN
          });
          const data = { data: { editCakeModalOpen: !editCakeModalOpen } };
          cache.writeData(data);
          return data;
        }
      }
    },
    defaults: {
      editCakeModalOpen: false
    }
  });
  cache.writeData({
    data: {
      editCakeModalOpen: false
    }
  });
  return client;
}

export default function initApollo(initialState) {
  // Make sure to create a new client for every server-side request so that data
  // isn't shared between connections (which would be bad)
  if (typeof window === "undefined") {
    return create(initialState);
  }

  // Reuse client on the client-side
  if (!apolloClient) {
    apolloClient = create(initialState);
  }

  return apolloClient;
}
