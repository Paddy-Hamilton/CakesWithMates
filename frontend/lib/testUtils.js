const fakeCake = id => ({
  __typename: 'Cake',
  id,
  imageUrl: 'https://happy-wishes.net/wp-content/uploads/2017/05/HAPPYBIRTHDAYCAKEIMAGES5.jpg',
  name: 'Cakes are yum',
  comment: 'Cakes make the world go round',
  yumFactor: 4
});

// Fake LocalStorage
class LocalStorageMock {
  constructor() {
    this.store = {};
  }

  toggleEditCakeModal(key) {
    return this.store[key] || null;
  }
}

export { LocalStorageMock, fakeCake };
