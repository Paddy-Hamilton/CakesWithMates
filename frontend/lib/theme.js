import { createMuiTheme } from "@material-ui/core/styles";
import purple from "@material-ui/core/colors/purple";
import teal from "@material-ui/core/colors/teal";
import green from "@material-ui/core/colors/green";

export const theme = createMuiTheme({
  palette: {
    primary: {
      light: purple[300],
      main: purple[500],
      dark: purple[700]
    },
    secondary: {
      xlight: teal["A100"],
      light: teal["A200"],
      main: teal["A400"],
      dark: teal["A700"]
    },
    success: green["600"]
  }
});