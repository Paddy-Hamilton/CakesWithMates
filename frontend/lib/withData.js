import withApollo from 'next-with-apollo';
import ApolloClient from 'apollo-boost';
import { EDIT_CAKE_MODAL_OPEN } from '../graphql/queries.graphql';
// can also be a function that accepts a `headers` object (SSR only) and returns a config

function createClient({ headers }) {
  return new ApolloClient({
    uri:
      process.env.NODE_ENV === 'development' ? 'http://localhost:8000/graphql' : 'https://ph-cake-api.now.sh/graphql',
    connectToDevTools: true, // Disables forceFetch on the server (so queries are only run once)
    request: async operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include'
        },
        headers
      });
    },
    clientState: {
      resolvers: {
        Mutation: {
          toggleEditCakeModal(_, variables, { cache }) {
            const { editCakeModalOpen } = cache.readQuery({ query: EDIT_CAKE_MODAL_OPEN });
            const data = { data: { editCakeModalOpen: !editCakeModalOpen } };
            cache.writeData(data);
            return data;
          }
        }
      },
      defaults: {
        editCakeModalOpen: false
      }
    }
  });
}

export default withApollo(createClient);
