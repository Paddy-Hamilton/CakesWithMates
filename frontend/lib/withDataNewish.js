import withApollo from "next-with-apollo";
import { ApolloClient } from "apollo-client";
import { InMemoryCache } from "apollo-cache-inmemory";
import { HttpLink } from "apollo-link-http";
import { onError } from "apollo-link-error";
import { withClientState } from "apollo-link-state";
import { ApolloLink, Observable } from "apollo-link";
import { EDIT_CAKE_MODAL_OPEN } from "../graphql/queries.graphql";
// can also be a function that accepts a `headers` object (SSR only) and returns a config

function createClient({ ctx, headers, initialState }) {
  const cache = new InMemoryCache();

  const request = async operation => {
    operation.setContext({
      fetchOptions: {
        credentials: "include"
      },
      headers
    });
  };

  const requestLink = new ApolloLink(
    (operation, forward) =>
      new Observable(observer => {
        let handle;
        Promise.resolve(operation)
          .then(oper => request(oper))
          .then(() => {
            handle = forward(operation).subscribe({
              next: observer.next.bind(observer),
              error: observer.error.bind(observer),
              complete: observer.complete.bind(observer)
            });
          })
          .catch(observer.error.bind(observer));

        return () => {
          if (handle) handle.unsubscribe();
        };
      })
  );

  const client = new ApolloClient({
    link: ApolloLink.from([
      onError(({ graphQLErrors, networkError }) => {
        if (graphQLErrors) {
          console.error(graphQLErrors);
        }
      }),
      requestLink,
      withClientState({
        defaults: {
          editCakeModalOpen: false
        },
        resolvers: {
          Mutation: {
            toggleEditCakeModal(_, variables, { cache }) {
              const { editCakeModalOpen } = cache.readQuery({
                query: EDIT_CAKE_MODAL_OPEN
              });
              const data = { data: { editCakeModalOpen: !editCakeModalOpen } };
              cache.writeData(data);
              return data;
            }
          }
        },
 \
      }),
      new HttpLink({
        uri:
          process.env.NODE_ENV === "development"
            ? "http://localhost:8000/graphql"
            : "https://ph-cake-api.now.sh/graphql",
        credentials: "include"
      })
    ]),
    connectToDevTools: true,
    ssrForceFetchDelay: 200
  });
  return client;
}

export default withApollo(createClient);
