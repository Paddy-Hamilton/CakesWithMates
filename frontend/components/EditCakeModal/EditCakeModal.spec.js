import React from 'react';
import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { fakeCake } from '../../lib/testUtils';
import wait from 'waait';
import { EditCakeModalStyled } from './index';
import { createMount } from '@material-ui/core/test-utils';

const cake = fakeCake('ASD123');
const defaultProps = { isOpen: { data: { editCakeModalOpen: true } } };

describe('<EditCakeModalStyled/>', () => {
  test('renders', () => {
    let mmount = createMount();
    const wrapper = mmount(<EditCakeModalStyled {...defaultProps} />);
    const h2 = wrapper.find('h2');
    expect(h2.text()).toBe('Create cake');
    expect(toJSON(h2)).toMatchSnapshot();
  });
  describe('When id is sent in props', () => {
    it('the form is prepopulated form', async () => {
      const wrapper = mount(<EditCakeModalStyled {...defaultProps} {...cake} />);

      const nameInput = wrapper.find('input#cake-name');
      const commentInput = wrapper.find('textarea#cake-comment');
      const imgUrlInput = wrapper.find('input#cake-image-url');
      expect(nameInput.get(0).props.value).toBe(cake.name);
      expect(commentInput.get(0).props.value).toBe(cake.comment);
      expect(imgUrlInput.get(0).props.value).toBe(cake.imageUrl);
    });
    it('the title is edit mode specific', () => {
      const wrapper = mount(<EditCakeModalStyled {...cake} {...defaultProps} />);
      const h2 = wrapper.find('h2');
      expect(h2.text()).toBe('Edit cake');
    });
  });
  it('displays changes', () => {
    const wrapper = mount(<EditCakeModalStyled {...defaultProps} />);
    const nameInput = wrapper.find('input#cake-name');
    nameInput.simulate('change', { target: { value: 'Simulated cake name' } });
    expect(wrapper.find('EditCakeModal').instance().state.name).toBe('Simulated cake name');
  });
});
