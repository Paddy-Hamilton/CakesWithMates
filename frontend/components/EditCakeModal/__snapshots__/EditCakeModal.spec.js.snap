// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditCakeModalStyled/> renders 1`] = `
<h2
  className="MuiTypography-root-8 MuiTypography-title-14 MuiTypography-gutterBottom-25 title"
>
  Create cake
</h2>
`;
